src/storage/src/compaction/picker.rs

```diff
@@ -12,28 +12,54 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use common_telemetry::debug;
+
 use crate::compaction::scheduler::CompactionRequestImpl;
+use crate::compaction::strategy::StrategyRef;
 use crate::compaction::task::{CompactionTask, CompactionTaskImpl};
 
-/// Picker picks input SST files and build the compaction task.
+/// Picker picks input SST files and builds the compaction task.
 /// Different compaction strategy may implement different pickers.
 pub trait Picker<R, T: CompactionTask>: Send + 'static {
-    fn pick(&self, req: &R) -> crate::error::Result<T>;
+    fn pick(&self, ctx: &PickerContext, req: &R) -> crate::error::Result<Option<T>>;
 }
 
+pub struct PickerContext {}
+
 /// L0 -> L1 all-to-all compaction based on time windows.
-pub(crate) struct SimplePicker {}
+pub(crate) struct SimplePicker {
+    strategy: StrategyRef,
+}
 
 #[allow(unused)]
 impl SimplePicker {
-    pub fn new() -> Self {
-        Self {}
+    pub fn new(strategy: StrategyRef) -> Self {
+        Self { strategy }
     }
 }
 
 impl Picker<CompactionRequestImpl, CompactionTaskImpl> for SimplePicker {
-    fn pick(&self, _req: &CompactionRequestImpl) -> crate::error::Result<CompactionTaskImpl> {
-        todo!()
+    fn pick(
+        &self,
+        ctx: &PickerContext,
+        req: &CompactionRequestImpl,
+    ) -> crate::error::Result<Option<CompactionTaskImpl>> {
+        let levels = req.levels();
+
+        for level_num in 0..levels.level_num() {
```

Comment:
It seems like the picker chains all the levels from L0 to the required level, but sometimes you may only want to compaction some levels. If the max level is 7, the compaction may happen only between 5-6 and left 0-4 unchanged.

Reply:
> It seems like the picker chains all the levels from L0 to the required level, but sometimes you may only want to compaction some levels. If the max level is 7, the compaction may happen only between 5-6 and left 0-4 unchanged.

A comprehension strategy should first calculate a score for each level and find the level with highest score for compaction, like RocksDB does:
https://github.com/facebook/rocksdb/blob/49628c9a83d337ea7b01f7ad3bdffa340e013913/db/compaction/compaction_picker_level.cc#L173

But here for simplicity, we just start every compaction from level 0 and returns as soon as we find the first "compactable" level, because for an ideal time-series workload, lower levels always contain the "hotter" data.